aws/cloudformation/db_parameters.yml.erb

```diff
@@ -0,0 +1,139 @@
+---
+# Database parameter group template
+# Reference documentation:
+# https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html
+# https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html
+# https://dev.mysql.com/doc/refman/5.7/en/replication-options-slave.html
+# https://dev.mysql.com/doc/refman/5.7/en/replication-options-binary-log.html
+
+# System variables used by primary RDS instance.
+Primary: &primary
+  # Enable Performance Schema.
+  performance_schema: 1
+
+  # Server System Variables:
+  # https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html
+
+  # Number of equality ranges in an equality comparison condition when the optimizer should switch from using index dives
+  # to index statistics in estimating the number of qualifying rows.
+  # Preserve 5.6 default of 10 (in 5.7 the default increased to 200).
+  eq_range_index_dive_limit: 10
+  # Store query logs in files to avoid database overhead.
+  log_output: FILE
+  # Set global execution timeout for SELECT statements, in milliseconds.
+  max_execution_time: 30000
+  # Skip resolving DNS on hostnames for improved performance.
+  skip_name_resolve: 1 # not dynamic
+  # Enable slow query log.
+  slow_query_log: 1
+  # Fully cache all table definitions.
+  table_definition_cache: 20000
+  # Set table_open_cache equal to max_connections.
+  # Ref: https://dev.mysql.com/doc/refman/5.7/en/table-cache.html
+  table_open_cache: '{DBInstanceClassMemory/12582880}'
```

Comment:
> Any concerns about the file descriptors limit mentioned in the linked docs?

Yes, potential conflicts with OS file-descriptor limits are worth keeping in mind (if I recall the default open files limit on RDS instances is currently 64k, which I think sets a soft limit on `(max_connections + table_open_cache*2)`, or about 20k each). I don't think we need to be too concerned about setting these parameters higher than the OS can handle since mysql auto-sizes the variable down if the file-descriptor limit isn't high enough to accommodate it, though we should keep in mind this auto-sizing might happen.

> More generally, I've been wondering about the max connections setting. If we think that having too many idle connections was causing a problem, and we were actually able to serve our peak traffic at a much lower number of connections when the DB was healthy, shouldn't we set this number lower? (Assuming that the application fails safely when it fails to create a connection)

This is a good idea- I'm not sure the application does fail safely when max-connections is reached (we should test that assumption in more detail if we planned to depend on it), but if it is safe we could put a hard cap on `max_connections` (using the `GREATEST` [DB Parameter Function](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_WorkingWithParamGroups.html#USER_ParamFunctions)) like `GREATEST({DBInstanceClassMemory/12582880}, 20000)`, or whatever the scaling threshold might be beyond which performance starts to degrade. I'll leave that potential tweak to a future follow-up though, since it could use more testing on how our application behaves when the limit is reached.

> May be moot if we're planning to switch to using a connection proxy like you'd mentioned.

I also don't expect us to need to scale back up to a large DB with 100+ frontends again until next December, so we have plenty of time to work on adding a connection proxy (or migrate to a database with a scalable thread-pool implementation) in the meantime without needing to worry too much about hitting these upper connection limits.